### PR TITLE
LPS-113196 Avoid two dispatch in a row when there is no valid connection

### PR DESCRIPTION
**Pull request by dgarciasarai**

### CODE REVIEW

**Comment by dgarciasarai** [comment]
ci:test:relevant

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 1]
```diff
@@ -164,61 +163,59 @@ export default function Chart({
...
+						actions.addDataSetItems({
 							dataSetItem: {histogram: [], value: null},
-							key,
+							dataSetItems,
+							keys,
 							timeSpanComparator,
 						});
-					});
+					}
+				}
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:210]: Awesome, overall I'd say this looks cleaner than before.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/state/chartState.js [hunk 2]
```diff
@@ -77,12 +77,22 @@ function reducer(state, action) {
...
+			nextState = action.payload.keys.reduce((state, key) => {
+				return addDataSetItem(
+					state,
+					{
+						...action.payload,
+						dataSetItem:
+							action.payload.dataSetItems &&
+							action.payload.dataSetItems[key] !== undefined
+								? action.payload.dataSetItems[key]
+								: action.payload.dataSetItem,
```
Comment (wincent) [inline modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/state/chartState.js:90]: It's a bit of a smell that the payload has singular and plural variants of the property:

- `dataSetItems`
- `dataSetItem`

Especially, in an action that is plural in nature (`ADD_DATA_SET_ITEMS`). It seems like `dataSetItem` is only used as a fallback, and it always has a fixed value of`{histogram: [], value: null}`.

So, maybe we should change this to:

```js
// Somewhere at the top of the module:
const FALLBACK_DATA_SET_ITEM = {histogram: [], value: null};

// In the reducer:
nextState = action.payload.keys.reduce((state, key) => {
	const dataSetItem =
		action.payload.dataSetItems &&
		action.payload.dataSetItems[key] !== undefined
			? action.payload.dataSetItems[key]
			: FALLBACK_DATA_SET_ITEM;

	return addDataSetItem(state, {
		...action.payload,
		dataSetItem,

		key,
	});
});
```

And the clean up the call-sites.

That ternary will get nicer when we turn on optional chaining and nullish coalescing; will be something like:

```js
const dataSetItem =
	action.payload.dataSetItems?.[key] ?? FALLBACK_DATA_SET_ITEM;
```

Created https://github.com/liferay/liferay-npm-tools/issues/450 to track turning that on.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 3]
```diff
@@ -164,61 +163,59 @@ export default function Chart({
...
-						let key = '';
+				});
 
-						if (getter.name === 'getHistoricalReads') {
-							key = 'analyticsReportsHistoricalReads';
-						}
-						else if (getter.name === 'getHistoricalViews') {
-							key = 'analyticsReportsHistoricalViews';
-						}
+				promises.push(promise);
```
[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:181]: I can do this instead:
```js
const promises = dataProviders.map((getter) => {
	return getter({
		timeSpanKey: chartState.timeSpanOption,
		timeSpanOffset: chartState.timeSpanOffset,
	});
});
```

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 4]
```diff
@@ -164,61 +163,59 @@ export default function Chart({
...
+
+						for (var i = 0; i < data.length; i++) {
+							if (data[i].status === 'fulfilled') {
+								dataSetItems = {
+									...dataSetItems,
+									...data[i].value,
+								};
+							}
+							else {
+								if (!hasHistoricalWarning) {
```
[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:197]: Of course! Thanks!

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 1 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:210]: ❤️ Thanks.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/state/chartState.js (hunk 2 above)
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/state/chartState.js:90]: Thanks. I'll change it and use `FALLBACK_DATA_SET_ITEM`.

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 5]
```diff
@@ -164,61 +163,59 @@ export default function Chart({
...
-							key = 'analyticsReportsHistoricalViews';
-						}
+				promises.push(promise);
+			});
 
-						if (!hasHistoricalWarning) {
-							addHistoricalWarning();
+			Promise.allSettled(promises).then((data) => {
+				if (!gone) {
+					if (isMounted()) {
```
[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:186]: 👍

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js [hunk 6]
```diff
@@ -164,61 +163,59 @@ export default function Chart({
...
-						}
-						else if (getter.name === 'getHistoricalViews') {
-							key = 'analyticsReportsHistoricalViews';
-						}
+				promises.push(promise);
+			});
 
-						if (!hasHistoricalWarning) {
-							addHistoricalWarning();
+			Promise.allSettled(promises).then((data) => {
```
[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:184]: 🍀

[the remark this answers is not in the document]
Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:184]: I think you linked the wrong URL...

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by dgarciasarai** [comment]
Everything works as expected in Chrome and Firefox. Let's see in IE11...

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by dgarciasarai** [comment]
ci:forward

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (wincent) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/components/Chart.js:180]: Meant to link: https://github.com/zloirock/core-js/pull/325

**Comment by dgarciasarai** [comment]
ci:test:sf

**Comment by dgarciasarai** [comment]
ci:test:relevant

**Comment by wincent** [comment]
lg2m to `ci:forward` if tests pass.

**Comment by dgarciasarai** [comment]
ci:stop

**Comment by dgarciasarai** [comment]
ci:forward

File: modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/state/chartState.js (hunk 2 above)
Comment (wincent) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/state/chartState.js]: FYI @dgarciasarai: support for `?.` and `??` [just landed](https://github.com/brianchandotcom/liferay-portal/pull/88602) in `master`, if you want to use it.

Comment (dgarciasarai) [reply modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/state/chartState.js]: I'll send a follow up pull request :) Thanks

**Pull request closed without merging**